src/analysis/processing/qgsalgorithmzonalstatistics.cpp

```diff
@@ -0,0 +1,153 @@
+/***************************************************************************
+                         qgsalgorithmzonalstatistics.cpp
+                         ---------------------
+    begin                : December 2019
+    copyright            : (C) 2019 by Alexander Bruy
+    email                : alexander dot bruy at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsalgorithmzonalstatistics.h"
+
+///@cond PRIVATE
+
+const std::vector< QgsZonalStatistics::Statistic > STATS
+{
+  QgsZonalStatistics::Count,
+  QgsZonalStatistics::Sum,
+  QgsZonalStatistics::Mean,
+  QgsZonalStatistics::Median,
+  QgsZonalStatistics::StDev,
+  QgsZonalStatistics::Min,
+  QgsZonalStatistics::Max,
+  QgsZonalStatistics::Range,
+  QgsZonalStatistics::Minority,
+  QgsZonalStatistics::Majority,
+  QgsZonalStatistics::Variety,
+  QgsZonalStatistics::Variance,
+};
+
+QString QgsZonalStatisticsAlgorithm::name() const
+{
+  return QStringLiteral( "zonalstatistics" );
+}
+
+QString QgsZonalStatisticsAlgorithm::displayName() const
+{
+  return QObject::tr( "Zonal statistics" );
+}
+
+QStringList QgsZonalStatisticsAlgorithm::tags() const
+{
+  return QObject::tr( "stats,statistics,zones,layer,sum,maximum,minimum,mean,count,standard,deviation,"
+                      "median,range,majority,minority,variety,variance,summary,raster" ).split( ',' );
+}
+
+QString QgsZonalStatisticsAlgorithm::group() const
+{
+  return QObject::tr( "Raster analysis" );
+}
+
+QString QgsZonalStatisticsAlgorithm::groupId() const
+{
+  return QStringLiteral( "rasteranalysis" );
+}
+
+QString QgsZonalStatisticsAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm calculates statistics of a raster layer for each feature "
+                      "of an overlapping polygon vector layer." );
+}
+
+QgsProcessingAlgorithm::Flags QgsZonalStatisticsAlgorithm::flags() const
+{
+  return QgsProcessingAlgorithm::flags() | QgsProcessingAlgorithm::FlagNoThreading;
```

Comment:
I don't suggest doing it in this PR -- but what do you think about us deprecating this alg, and adding a new one which doesn't modify layers in place? I think it's more in line with how processing works...

Reply:
Sounds like a plan.

Reply:
Do you think we should also update `QgsZonalStatistics` class to produce new layer?

Reply:
Yes please.